sql: rewrite zone configs and partitions on DROP INDEX failure for RBR

otan opened the pull request:
See individual commits for details.

Refs: #63974

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/63976)
<!-- Reviewable:end -->

arulajmani reviewed (changes requested):
Mostly nits, but this looks promising :tada: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63976#-:-MYrA_-q17EE1Hj2fbMK:b6p8tcc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/ccl/multiregionccl/region_test.go, line 428 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63976#-MYr7NBmEHqHsT38nVOA:-MYr7NBmEHqHsT38nVOB:bgp8wi9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2497ed6350e2e8330ea52731fe198259a428dbe/pkg/ccl/multiregionccl/region_test.go#L428)):*
> ```Go
> 			wg.Wait()
> 
> 			require.NoError(
> ```

You might need a succeedsSoon here, because the `OnFailOrCancel` job runs async, right?
___
*[pkg/ccl/multiregionccl/region_test.go, line 608 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63976#-MYr6klp8aVBWLmp9Cs3:-MYr6klq6SJGc_97Khah:bk9rd2q) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2497ed6350e2e8330ea52731fe198259a428dbe/pkg/ccl/multiregionccl/region_test.go#L608)):*
> ```Go
> }
> 
> func validateIndexAndPartitions(
> ```

Let's move this to `testutils`?
___
*[pkg/sql/create_index.go, line 562 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63976#-MYrAGATBYmXUH-HTTCB:-MYrAGATBYmXUH-HTTCC:bqj027b) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2497ed6350e2e8330ea52731fe198259a428dbe/pkg/sql/create_index.go#L562)):*
> ```Go
> func (*createIndexNode) Close(context.Context)        {}
> 
> func (p *planner) configureIndexDescForNewIndexPartitioning(
> ```

Same as below, can we get rid of this method and just make all callers call the function?
___
*[pkg/sql/create_index.go, line 626 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63976#-MYrA3Pw8Jk3YsdV6-0O:-MYrA3Pw8Jk3YsdV6-0P:bs4whsj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2497ed6350e2e8330ea52731fe198259a428dbe/pkg/sql/create_index.go#L626)):*
> ```Go
> }
> 
> func (p *planner) configureZoneConfigForNewIndexPartitioning(
> ```

This function feels redundant now, can we just get rid of it?
___
*[pkg/sql/create_index.go, line 641 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63976#-MYr8jNE7Hu37qFn35hd:-MYr8jNE7Hu37qFn35he:b-f9lml7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2497ed6350e2e8330ea52731fe198259a428dbe/pkg/sql/create_index.go#L641)):*
> ```Go
> // configureZoneConfigForNewIndexPartitioning configures the zone config for any new index
> // in a REGIONAL BY ROW table.
> // This *must* be done after the index ID has been allocated.	func configureZoneConfigForNewIndexPartitioning(
> ```

nit: unintentional after the full stop? 
___
*[pkg/sql/schema_changer.go, line 2470 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63976#-MYr8JNFBuc8iN5_61Mp:-MYr8JNFBuc8iN5_61Mq:b-v4yrks) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2497ed6350e2e8330ea52731fe198259a428dbe/pkg/sql/schema_changer.go#L2470)):*
> ```Go
> 
> 	if pkSwap := mutation.AsPrimaryKeySwap(); pkSwap != nil {
> 		// TODO(#63974): we should re-write the partitions for REGIONAL BY ROW too.
> ```

Maybe point the issue to https://github.com/cockroachdb/cockroach/issues/64011?


<!-- Sent from Reviewable.io -->

otan committed:
sql: rewrite zone configs and partitions on DROP INDEX failure for RBR

Release note (bug fix): Previously, a DROP INDEX failing when a concurrent
DROP REGION succeeds would leave the database in a irrecoverable state.
This has now been resolved.

arulajmani reviewed (commented):
This diff will make you happy, atleast for getting the scope of this PR. 

```
@@ -2430,7 +2430,15 @@ func (sc *SchemaChanger) applyPartitionAndZoneConfigChangeForMutation(
        // REGIONAL BY ROW tables need to have CREATE or DROP INDEX repartition
        // just before finalization, in case an ADD/DROP REGION races with it.
        if tableDesc.IsLocalityRegionalByRow() {
-               if indexMut := mutation.AsIndex(); indexMut != nil {
+               pkSwapID := descpb.InvalidMutationID
+               // TODO(arul): This is a hack. 
+               for _, m := range tableDesc.AllMutations() {
+                       if m.AsPrimaryKeySwap() != nil {
+                               pkSwapID = m.MutationID()
+                               break
+                       }
+               }
+               if indexMut := mutation.AsIndex(); indexMut != nil && indexMut.MutationID() != pkSwapID {
```

Altering to/from RBR and alter primary key operations on RBR tables requires a little bit of restructuring here, but the idea is the same -- we have to correlate index mutations to whether they're queued up by an ALTER PRIMARY KEY/ALTER TABLE LOCALITY TO RBR/ALTER TABLE LOCALITY FROM RBR and use the information to negotiate how to repartition.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63976#-:-MYuYox_41_N2oIM1nTr:b-e6b83n)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)



<!-- Sent from Reviewable.io -->

otan commented:
DIE DIE DIE
